Add non-zero length style option to `explicit-length-check` rule

paulobmarcos opened the pull request:
Fixes #56

In short it aims to add functionality to the rule `explicit-length-check` by providing some options to determine the preferred way of checking for length properties.

The `explicit-length-check` rule now enforces length comparison with `!== 0` when checking for an empty array. It auto-fixes when the check is made with '< 1'. 

#### Fail
```js
if (string.length < 1) {}
```
#### Pass
```js
if (array.length !== 0) {}
```

For non zero length comparisons we can define our preferred way of checking it by providing an option:
```js
{
  'unicorn/explicit-length-check': ['error', {
    'non-zero': 'not-equal'
  }]
}
```
where `non-zero` can be one of the following:
- `not-equal`
	- this option makes sure that non-zero is checked with: `array.length !== 0`
- `greater-than`
	- this option makes sure that non-zero is checked with: `array.length > 0`
- `greater-than-or-equal`
	- this option makes sure that non-zero is checked with: `array.length >= 1`

*If no options are provided it will work as before, allowing any of this possibilities.* 


What I did was to check for `BinaryExpressions` and look for the `empty` case and, if defined in the options, the `not-empty` case.

It will check if any of the cases, `empty` or `not-empty`, are contemplated in the `BinaryExpression`  by comparing the operator and the literal value. 
If any of the cases have a match and the operator+value are of other option types, than it is rejected and a fix is provided with the preferred way.

sindresorhus commented:
Thanks for working on this :)

I think we should just drop the `emptyOption` option. `a.length < 1` makes no sense` and I've never seen it in the wild.

I'm ok with the `notEmptyOption` option.

sindresorhus commented:
> Also added a new check. When the literal comes before the length property it also rejects it.
1 > array.length should be array.length < 1. What do you think?

That should already be handled by the ESLInt [`yoda`](http://eslint.org/docs/rules/yoda) rule.

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -1,5 +1,93 @@
+function checkEmptyType(type, operator, value) {
+	switch (type) {
+		case 'eq':
+			if (types.lt.includes(operator) && value === 1) {
```
You can't use `.includes()`. We still support Node.js 4.

paulobmarcos replied:
Yes I noticed that and pushed a fix for it right after 😁

sindresorhus commented:
Would you also be able to add a fixer? So it can be auto-fixed.

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -1,5 +1,93 @@
+			if ((types.gte.includes(operator) && value === 1) ||
+				(types.ne.includes(operator) && value === 0)
+			) {
+				return 'not empty `length` should be compared with `> 0`.';
```
```js
return 'Non-zero `.length` should be compared with `> 0`.';
```

paulobmarcos replied:
Good one! It reads better that way. ✌️

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -1,5 +1,93 @@
+function checkEmptyType(type, operator, value) {
+	switch (type) {
+		case 'eq':
+			if (types.lt.indexOf(operator) >= 0 && value === 1) {
```
`>= 0` → `!== -1`

paulobmarcos commented:
Hey @sindresorhus! Sorry, didn't have the opportunity to look into this right away.
I removed the `emptyOption` and it now check for `!== 0` as the default. 
Also added an auto-fixer for both empty and non zero length comparisons. 
And you are right, the ESLint `yoda` rule already handles one of the cases 👍 

Hope it looks good 🎉

sindresorhus commented on docs/rules/explicit-length-check.md:
````diff
@@ -11,11 +11,46 @@ if (!array.length) {}
+```js
+{
+  "unicorn/explicit-length-check": ["error", {
+    "not-empty": "notEmptyOption"
````
I think `not-empty` should be `non-zero`.

And use single-quotes.

paulobmarcos replied:
`non-zero` seems good 👍 

One thing, since this configuration is done in a `.json` file and since the JSON standard is double-quotes, shouldn't the example also be in double-quotes?

sindresorhus replied:
ESLint config can be in both JS and JSON.

paulobmarcos replied:
Ok, I will change it to single-quotes!

sindresorhus commented on docs/rules/explicit-length-check.md:
````diff
@@ -11,11 +11,46 @@ if (!array.length) {}
+if (array.length !== 0) {}
+```
+
+## Non Zero Comparisons
````
Non-Zero

sindresorhus commented on docs/rules/explicit-length-check.md:
```diff
@@ -11,11 +11,46 @@ if (!array.length) {}
+- `"gt"` (greater-than)
+	- this option makes sure that non zero is checked with: `array.length > 0`
+- `"gte"` (greater-than or equal-to)
+	- this option makes sure that non zero is checked with: `array.length >= 1`
```
I would prefer to use verbose option values for readability. So:

- `not-equal`
- `greater-than`
- `greater-than-or-equal`

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -1,5 +1,95 @@
+		operator,
+		value
+	};
+};
```
This feels kinda moot. Just use an object in the actual calls instead.

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -33,5 +125,7 @@ const create = context => {
 
 module.exports = {
 	create,
-	meta: {}
```
The option should be defined using a `schema` so it can be validated by ESLint

See: http://eslint.org/docs/developer-guide/working-with-rules#options-schemas

sindresorhus commented on rules/explicit-length-check.js:
```diff
@@ -1,5 +1,95 @@
+	if (node.operator === '<' && node.right.value === 1) {
+		reportError(context,
+			node,
+			'Empty `.length` should be compared with `=== 0`.',
```
Empty => Zero

sindresorhus commented:
Sorry about the slow response. Summer ;)

Looks pretty good. Just some last minor nitpick.

paulobmarcos commented:
Hey @sindresorhus! No worries!

Just checked your review and pushed some changes. ✌️

sindresorhus commented:
Great, thanks :)
